lib/obj.js

```diff
@@ -1,365 +1,479 @@
-        });
+    }).then(function() {
+        return materials;
     });
 }
 
-function getImages(inputPath, materials) {
-    // Collect all the image files from the materials
-    var images = [];
+function loadImages(imagePaths, objPath, options) {
+    var images = {};
+    return Promise.map(imagePaths, function(imagePath) {
+        if (options.secure && outsideDirectory(imagePath, objPath)) {
+            options.logger('Could not read image file at ' + imagePath + ' because it is outside of the obj directory and the secure flag is true. Material will ignore this image.');
+            return;
+        }
+        return loadImage(imagePath, options)
+            .then(function(image) {
+                if (defined(image)) {
+                    images[imagePath] = image;
+                }
+            })
+            .catch(function() {
+                options.logger('Could not read image file at ' + imagePath + '. Material will ignore this image.');
+                return undefined;
+            });
+    }).then(function() {
+        return images;
+    });
+}
+
+function getImagePaths(materials) {
+    var imagePaths = {};
     for (var name in materials) {
         if (materials.hasOwnProperty(name)) {
             var material = materials[name];
-            if (defined(material.ambientColorMap) && (images.indexOf(material.ambientColorMap) === -1)) {
-                images.push(material.ambientColorMap);
+            if (defined(material.ambientColorMap)) {
+                imagePaths[material.ambientColorMap] = true;
             }
-            if (defined(material.diffuseColorMap) && (images.indexOf(material.diffuseColorMap) === -1)) {
-                images.push(material.diffuseColorMap);
+            if (defined(material.diffuseColorMap)) {
+                imagePaths[material.diffuseColorMap] = true;
             }
-            if (defined(material.emissionColorMap) && (images.indexOf(material.emissionColorMap) === -1)) {
-                images.push(material.emissionColorMap);
+            if (defined(material.emissionColorMap)) {
+                imagePaths[material.emissionColorMap] = true;
             }
-            if (defined(material.specularColorMap) && (images.indexOf(material.specularColorMap) === -1)) {
-                images.push(material.specularColorMap);
+            if (defined(material.specularColorMap)) {
+                imagePaths[material.specularColorMap] = true;
             }
         }
     }
+    return Object.keys(imagePaths);
+}
 
-    // Load the image files
-    var promises = [];
-    var imagesInfo = {};
-    var imagesLength = images.length;
-    for (var i = 0; i < imagesLength; i++) {
-        var imagePath = images[i];
-        if (!path.isAbsolute(imagePath)) {
-            imagePath = path.join(inputPath, imagePath);
+function removeEmptyPrimitives(primitives) {
+    var final = [];
+    var primitivesLength = primitives.length;
+    for (var i = 0; i < primitivesLength; ++i) {
```

Comment:
This function could be simplified into a single `primitives.filter` call (and since it's only called once, we probably don't even need a separate function for this).